File: apps/lorepia/src/lib/ipc/contracts/generation.ts
import type { AssetDeliveryDto } from './character';

import type {
    CreatorControlValue,
    OrchestrationConditionExprDto,
    OrchestrationVariableMapDto,
    OrchestrationVariableRefDto,
    OrchestrationVariableValueDto,
    PromptBlockKind,
    PromptBlockOverflowPolicy,
    PromptBlockRoleHint,
    SafePromptTemplateDto,
} from './common';

import type { ConversationBranchDto, ConversationMode, MessageStatus } from './conversation';

import type { MemorySelectionLane, MemorySelectionReasonDto } from './memory';

export type InteractionUiRegionDto =
    'message' | 'background' | 'character_portrait' | 'status_panel' | 'audio';

export interface InteractionChoiceDto {
    id: string;
    label: string;
}

export type InteractionEffectProjectionRejectionReasonDto =
    'unsafe_native_text' | 'invalid_stored_effect' | 'asset_unavailable';

export type InteractionEffectDto =
    | { kind: 'state_changed' }
    | { kind: 'knowledge_activated'; entry_id: string }
    | { kind: 'show_asset'; asset: AssetDeliveryDto; region: InteractionUiRegionDto }
    | { kind: 'play_audio'; asset: AssetDeliveryDto }
    | { kind: 'present_choices'; choices: InteractionChoiceDto[] }
    | { kind: 'visible_system_event'; text: string }
    | {
          kind: 'dice_rolled';
          count: number;
          sides: number;
          modifier: number;
          rolls: number[];
          total: number;
      }
    | {
          kind: 'approval_pending';
          title: string;
          body: string;
          expires_after_seconds: number | null;
      }
    | {
          kind: 'projection_rejected';
          reason: InteractionEffectProjectionRejectionReasonDto;
      };

export interface InteractionEffectEventDto {
    delivery_id: string;
    effect_id: string;
    conversation_id: string;
    branch_id: string;
    resulting_state_revision: number;
    event_created_at: string;
    effect: InteractionEffectDto;
}

export interface DecideInteractionProposalInput {
    conversation_id: string;
    branch_id: string;
    proposal_record_id: string;
    expected_state_revision: number;
    expected_proposal_revision: number;
    decision: 'approve' | 'reject';
}

export interface InteractionProposalRecordDto {
    id: string;
    title: string;
    body: string;
    projection_rejection_reason?: 'unsafe_native_text';
    status: 'pending' | 'approved' | 'rejected' | 'expired';
    source_interaction_state_revision: number;
    requested_at_epoch_seconds: number;
    expires_at_epoch_seconds: number | null;
    decided_at_epoch_seconds: number | null;
}

export interface InteractionProposalDecisionReceiptDto {
    proposal: InteractionProposalRecordDto;
    state_revision: number;
    effects: InteractionEffectDto[];
}

export interface ListInteractionProposalsInput {
    conversation_id: string;
    branch_id: string;
    status: 'pending' | 'approved' | 'rejected' | 'expired';
    limit: number;
}

export interface InteractionProposalListItemDto {
    conversation_id: string;
    branch_id: string;
    state_revision: number;
    proposal_revision: number;
    proposal: InteractionProposalRecordDto;
}

export interface ExpireInteractionProposalsInput {
    conversation_id: string;
    branch_id: string;
    limit: number;
}

export interface InteractionProposalExpiryReceiptDto {
    conversation_id: string;
    branch_id: string;
    current_state_revision: number;
    expired_proposals: InteractionProposalListItemDto[];
    has_more_expired: boolean;
}

export interface GenerationAttemptProposalDto {
    id: string;
    title: string;
    body: string;
    projection_rejection_reason?: 'unsafe_native_text';
    status: 'pending' | 'approved' | 'rejected' | 'expired';
    /** Canonical u64 decimal (including zero); never coerce to a JavaScript number. */
    source_interaction_state_revision: string;
    requested_at_epoch_seconds: number;
    expires_at_epoch_seconds: number | null;
    decided_at_epoch_seconds: number | null;
}

export interface ListGenerationAttemptProposalsInput {
    conversation_id: string;
    source_branch_id: string;
    status: 'pending' | 'approved' | 'rejected' | 'expired';
    limit: number;
}

export interface GenerationAttemptProposalListItemDto {
    conversation_id: string;
    source_branch_id: string;
    proposed_branch_id: string;
    generation_id: string;
    /** Canonical positive u64 decimal; echo unchanged for decision CAS. */
    aggregate_revision: string;
    /** Canonical positive u64 decimal; presentation-only. */
    interaction_state_revision: string;
    pending_proposal_count: number;
    /** Canonical positive u64 decimal; echo unchanged for decision CAS. */
    proposal_revision: string;
    proposal: GenerationAttemptProposalDto;
}

export interface DecideGenerationAttemptProposalInput {
    conversation_id: string;
    source_branch_id: string;
    generation_id: string;
    proposal_record_id: string;
    expected_aggregate_revision: string;
    expected_proposal_revision: string;
    decision: 'approve' | 'reject';
}

export interface GenerationAttemptProposalDecisionReceiptDto {
    conversation_id: string;
    source_branch_id: string;
    proposed_branch_id: string;
    generation_id: string;
    aggregate_revision: string;
    interaction_state_revision: string;
    pending_proposal_count: number;
    proposal_revision: string;
    proposal: GenerationAttemptProposalDto;
    approval_evidence_sha256: string | null;
    exact_replay: boolean;
}

export interface ExpireGenerationAttemptProposalsInput {
    conversation_id: string;
    source_branch_id: string;
    limit: number;
}

export interface GenerationAttemptProposalExpiryReceiptDto {
    conversation_id: string;
    source_branch_id: string;
    decisions: GenerationAttemptProposalDecisionReceiptDto[];
    has_more_due: boolean;
}

export type RetryableGenerationAttemptStatusDto = 'before_generation_applied' | 'dispatch_ready';

export interface ListRetryableGenerationAttemptsInput {
    conversation_id: string;
    source_branch_id: string;
    limit: number;
}

/** Safe restart projection; operation, prompt, provider, and credential state stays in Rust. */
export interface RetryableGenerationAttemptDto {
    generation_id: string;
    status: RetryableGenerationAttemptStatusDto;
    created_at: string;
    updated_at: string;
}

export interface InteractionEffectHistoryCursorDto {
    resulting_state_revision: number;
    sequence: number;
}

export interface ListInteractionEffectHistoryInput {
    conversation_id: string;
    branch_id: string;
    after: InteractionEffectHistoryCursorDto | null;
    limit: number;
}

export type InteractionChoiceStatusDto = 'pending' | 'consumed' | 'expired';

export interface InteractionEffectHistoryItemDto {
    effect_id: string;
    conversation_id: string;
    branch_id: string;
    resulting_state_revision: number;
    sequence: number;
    event_created_at: string;
    replay_on_reopen: boolean;
    choice_status: InteractionChoiceStatusDto | null;
    selected_choice_id: string | null;
    choice_decided_at_epoch_seconds: number | null;
    effect: InteractionEffectDto;
}

export interface InteractionEffectHistoryPageDto {
    current_state_revision: number;
    items: InteractionEffectHistoryItemDto[];
    next_cursor: InteractionEffectHistoryCursorDto | null;
}

export interface ListReopenInteractionEffectsInput {
    conversation_id: string;
    branch_id: string;
    limit: number;
}

export interface InteractionReopenSnapshotDto {
    current_state_revision: number;
    items: InteractionEffectHistoryItemDto[];
    older_cursor: InteractionEffectHistoryCursorDto | null;
}

export interface SubmitInteractionChoiceInput {
    conversation_id: string;
    branch_id: string;
    effect_id: string;
    choice_id: string;
    expected_state_revision: number;
}

export interface InteractionChoiceSelectionReceiptDto {
    choice_effect: InteractionEffectHistoryItemDto;
    resulting_state_revision: number;
}

export interface RoomInteractionClientApi {
    expireInteractionProposals(
        input: ExpireInteractionProposalsInput,
    ): Promise<InteractionProposalExpiryReceiptDto>;
    listInteractionProposals(
        input: ListInteractionProposalsInput,
    ): Promise<InteractionProposalListItemDto[]>;
    listInteractionEffectHistory(
        input: ListInteractionEffectHistoryInput,
    ): Promise<InteractionEffectHistoryPageDto>;
    listReopenInteractionEffects(
        input: ListReopenInteractionEffectsInput,
    ): Promise<InteractionReopenSnapshotDto>;
    submitInteractionChoice(
        input: SubmitInteractionChoiceInput,
    ): Promise<InteractionChoiceSelectionReceiptDto>;
}

export interface GenerationAttemptApprovalClientApi {
    expireGenerationAttemptProposals(
        input: ExpireGenerationAttemptProposalsInput,
    ): Promise<GenerationAttemptProposalExpiryReceiptDto>;
    listGenerationAttemptProposals(
        input: ListGenerationAttemptProposalsInput,
    ): Promise<GenerationAttemptProposalListItemDto[]>;
    listRetryableGenerationAttempts(
        input: ListRetryableGenerationAttemptsInput,
    ): Promise<RetryableGenerationAttemptDto[]>;
    decideGenerationAttemptProposal(
        input: DecideGenerationAttemptProposalInput,
    ): Promise<GenerationAttemptProposalDecisionReceiptDto>;
}

export interface GenerationTargetDto {
    model_route_id: string;
    generation_preset_id: string;
}

export interface GenerationUsageDto {
    input_tokens: number | null;
    cached_read_tokens: number | null;
    cached_write_tokens: number | null;
    output_tokens: number | null;
    reasoning_tokens: number | null;
    tool_tokens: number | null;
}

export type ChatEventKindDto =
    | { type: 'generation_started' }
    | { type: 'reasoning_delta'; payload: string }
    | { type: 'text_delta'; payload: string }
    | { type: 'tool_call_started'; payload: { id: string; name: string } }
    | { type: 'tool_call_arguments_delta'; payload: { id: string; delta: string } }
    | { type: 'tool_call_completed'; payload: { id: string } }
    | { type: 'usage_updated'; payload: GenerationUsageDto }
    | { type: 'message_committed'; payload: { message_id: string; status: MessageStatus } }
    | { type: 'generation_cancelled' }
    | { type: 'generation_failed'; payload: { code: string; message: string } }
    | { type: 'generation_finished' };

export interface ChatEventDto {
    event_version: number;
    generation_id: string;
    conversation_id: string;
    branch_id: string | null;
    assistant_message_id: string | null;
    sequence: number;
    emitted_at: string;
    kind: ChatEventKindDto;
}

export type ChatStreamItemDto =
    | { type: 'event'; payload: ChatEventDto }
    | {
          type: 'reconciliation_required';
          payload: {
              reason:
                  | 'broadcast_lagged'
                  | 'unsupported_event_version'
                  | 'route_mismatch'
                  | 'duplicate_or_decreasing_sequence'
                  | 'sequence_gap'
                  | 'live_snapshot'
                  | 'event_after_terminal';
              generation_id: string;
              conversation_id: string;
              branch_id: string;
              last_sequence: number | null;
              observed_sequence: number | null;
              dropped_events: number | null;
              supported_event_version: number;
              display_prefix: string | null;
              reasoning_prefix: string | null;
          };
      }
    | { type: 'closed' };

export type GenerationSelectionInput =
    | { kind: 'legacy_profile'; provider_profile_id: string }
    | { kind: 'target'; target: GenerationTargetDto };

export type RuntimePromptRoleInput = 'system' | 'user' | 'assistant';

export interface RuntimePromptMessageInput {
    role: RuntimePromptRoleInput;
    content: string;
}

export interface GenerateRuntimeTextInput {
    request_id: string;
    audit: {
        character_id: string;
        character_content_revision_id: string | null;
        capability: 'model:primary' | 'model:auxiliary';
        grant_sha256: string;
    };
    selection: GenerationSelectionInput;
    messages: RuntimePromptMessageInput[];
}

export interface RuntimeTextGenerationDto {
    request_id: string;
    result: string;
    usage: GenerationUsageDto;
}

export interface SendMessageInput {
    conversation_id: string;
    branch_id: string;
    expected_head: string | null;
    mode: ConversationMode;
    text: string;
    selection: GenerationSelectionInput;
    /** Per-generation character/runtime values merged after stored prompt state. */
    variable_overrides?: OrchestrationVariableMapDto;
    /** New and resume identities are mutually exclusive; supply exactly one. */
    operation_nonce?: string | null;
    generation_attempt_id?: string | null;
}

export interface GenerationStartedDto {
    generation_id: string;
}

export interface MessageActionGenerationDto {
    branch: ConversationBranchDto;
    generation_id: string;
}

export interface EditUserMessageInput {
    conversation_id: string;
    branch_id: string;
    expected_head: string | null;
    message_id: string;
    replacement_text: string;
    selection: GenerationSelectionInput;
    /** New and resume identities are mutually exclusive; supply exactly one. */
    operation_nonce?: string | null;
    generation_attempt_id?: string | null;
}

export interface RegenerateAssistantMessageInput {
    conversation_id: string;
    branch_id: string;
    expected_head: string | null;
    message_id: string;
    selection: GenerationSelectionInput;
    /** New and resume identities are mutually exclusive; supply exactly one. */
    operation_nonce?: string | null;
    generation_attempt_id?: string | null;
}

export interface RemoveMessageInput {
    conversation_id: string;
    branch_id: string;
    expected_head: string | null;
    message_id: string;
}

export type CreatorValueExprDto =
    | { kind: 'literal'; value: OrchestrationVariableValueDto }
    | { kind: 'variable'; variable: OrchestrationVariableRefDto };

export type CreatorInteractionEventDto =
    | { kind: 'conversation_opened' }
    | { kind: 'conversation_started' }
    | { kind: 'before_generation' }
    | { kind: 'after_generation' }
    | { kind: 'message_committed' }
    | { kind: 'user_action'; action_id: string }
    | { kind: 'variable_changed'; variable: OrchestrationVariableRefDto }
    | { kind: 'knowledge_activated'; entry_id: string };

export interface CreatorInteractionChoiceDto {
    id: string;
    label: string;
    value: OrchestrationVariableValueDto;
    enabled_when: OrchestrationConditionExprDto | null;
}

export type CreatorInteractionActionDto =
    | {
          kind: 'set_variable';
          target: OrchestrationVariableRefDto;
          value: CreatorValueExprDto;
      }
    | { kind: 'increment_variable'; target: OrchestrationVariableRefDto; amount: number }
    | { kind: 'activate_knowledge'; entry_id: string }
    | {
          kind: 'show_asset';
          asset_id: string;
          region: 'message' | 'background' | 'character_portrait' | 'status_panel' | 'audio';
      }
    | { kind: 'play_audio'; asset_id: string }
    | { kind: 'present_choices'; choices: CreatorInteractionChoiceDto[] }
    | { kind: 'append_visible_system_event'; text: SafePromptTemplateDto }
    | {
          kind: 'roll_dice';
          expression: { count: number; sides: number; modifier: number };
          target: OrchestrationVariableRefDto | null;
      }
    | {
          kind: 'request_user_approval';
          proposal: {
              id: string;
              title: string;
              body: SafePromptTemplateDto;
              expires_after_seconds: number | null;
          };
      };

export interface CreatorInteractionRuleDocumentDto {
    id: string;
    name: string;
    enabled: boolean;
    event: CreatorInteractionEventDto;
    condition: OrchestrationConditionExprDto | null;
    actions: CreatorInteractionActionDto[];
    priority: number;
    stop_after_match: boolean;
}

export interface CreatorInteractionRuleSetDocumentDto {
    id: string;
    name: string;
    rules: CreatorInteractionRuleDocumentDto[];
    max_actions_per_event: number;
}

export interface UpsertInteractionRuleSetInput {
    value: CreatorInteractionRuleSetDocumentDto;
    expected_revision: number | null;
}

export interface GetInteractionRuleSetInput {
    interaction_rule_set_id: string;
}

export interface DeleteInteractionRuleSetInput extends GetInteractionRuleSetInput {
    expected_revision: number;
}

export interface InteractionStateEntryDto {
    id: string;
    label: string;
    value: CreatorControlValue;
    scope: string;
}

export interface PromptPlanMessagePreviewDto {
    sequence: number;
    block_id: string;
    block_kind: PromptBlockKind;
    requested_role: PromptBlockRoleHint;
    effective_role: 'system' | 'developer' | 'user' | 'assistant';
    estimated_tokens: number;
    source_message_ids: string[];
    truncated: boolean;
}

export type PromptKnowledgeSelectionReasonDto =
    | { kind: 'always' }
    | { kind: 'manual' }
    | { kind: 'keyword' }
    | { kind: 'regex' }
    | { kind: 'semantic'; score_millionths: number }
    | { kind: 'condition' }
    | { kind: 'recursive'; parent_id: string };

export type PromptEvidenceExclusionCodeDto =
    | 'entry_disabled'
    | 'activation_probability_gate'
    | 'activation_rule_did_not_match'
    | 'per_entry_token_limit'
    | 'knowledge_token_budget_overflow'
    | 'knowledge_remaining_token_budget'
    | 'prompt_token_budget'
    | 'memory_retrieval_count_limit'
    | 'memory_remaining_token_budget'
    | 'other_conversation'
    | 'memory_invalidated'
    | 'excluded_from_conversation'
    | 'excluded_from_character'
    | 'not_on_active_branch_lineage'
    | 'reversed_source_range'
    | 'other';

export interface PromptKnowledgeSelectionEvidenceDto {
    entry_id: string;
    selected: boolean;
    reasons: PromptKnowledgeSelectionReasonDto[];
    estimated_tokens: number;
    exclusion_code: PromptEvidenceExclusionCodeDto | null;
}

export interface PromptMemorySelectionEvidenceDto {
    record_id: string;
    selected: boolean;
    lane: MemorySelectionLane | null;
    rank_millionths: number | null;
    estimated_tokens: number;
    reasons: MemorySelectionReasonDto[];
    exclusion_code: PromptEvidenceExclusionCodeDto | null;
}

export interface PromptBlockSourceTraceDto {
    authority: 'application' | 'creator' | 'user' | 'conversation' | 'imported_content';
    source_kind:
        | 'application_built_in'
        | 'user_created'
        | 'imported_standard'
        | 'imported_package'
        | 'generated';
    source_id: string | null;
    source_revision: string | null;
    source_hash: string | null;
}

export interface PromptBlockResolutionTraceDto {
    block_id: string;
    block_kind: PromptBlockKind;
    source: PromptBlockSourceTraceDto;
    status:
        | 'included'
        | 'condition_false'
        | 'disabled'
        | 'empty'
        | 'dropped_for_budget'
        | 'trimmed_head'
        | 'trimmed_tail'
        | 'reduced_items'
        | 'summarized';
    original_estimated_tokens: number;
    final_estimated_tokens: number;
    produced_message_count: number;
    knowledge_evidence: PromptKnowledgeSelectionEvidenceDto[];
    memory_record_ids: string[];
    memory_evidence: PromptMemorySelectionEvidenceDto[];
    truncated: boolean;
}

export interface PromptRoleMappingTraceDto {
    block_id: string;
    requested_role: PromptBlockRoleHint;
    effective_role: 'system' | 'developer' | 'user' | 'assistant';
}

export interface PromptOverflowTraceDto {
    block_id: string;
    policy: PromptBlockOverflowPolicy;
    tokens_before: number;
    tokens_after: number;
}

export interface PromptCacheDirectivePreviewDto {
    boundary_id: string;
    after_block_id: string;
    after_message_sequence: number | null;
    role_filter: PromptCacheRoleFilterDto;
    ttl: PromptCacheTtl;
    mode: PromptCacheMode;
    status: 'applied' | 'ignored_unsupported' | 'ignored_limit' | 'removed_with_block';
}

export type PromptProviderFamily =
    | 'openai_responses'
    | 'openai_chat_completions'
    | 'anthropic_messages'
    | 'gemini_generate_content'
    | 'ollama_native';

export type PromptCacheRoleFilterDto =
    { kind: 'all' } | { kind: 'system_like' } | { kind: 'exact_role'; role: PromptBlockRoleHint };

export type PromptCacheTtl = 'provider_default' | 'short' | 'long';

export type PromptCacheMode = 'automatic' | 'explicit' | 'disabled';

export interface PromptProviderMessagePreviewDto {
    sequence: number;
    block_id: string;
    effective_role: 'system' | 'developer' | 'user' | 'assistant';
    wire_role: 'system' | 'developer' | 'user' | 'assistant' | 'model';
    placement: 'message' | 'system_instruction';
    estimated_tokens: number;
}

export type PromptProviderCacheBoundaryWarning =
    | 'prompt_caching_unavailable'
    | 'provider_managed_caching_has_no_explicit_boundary'
    | 'explicit_cached_context_must_be_created_separately'
    | 'requested_cache_mode_unavailable'
    | 'cache_disable_unavailable'
    | 'cache_boundary_limit_exceeded'
    | 'cache_boundary_target_was_removed'
    | 'cache_role_filter_unavailable'
    | 'requested_cache_ttl_unavailable';

export type PromptProviderCacheBoundaryDispositionDto =
    | { disposition: 'no_directive' }
    | {
          disposition: 'mapped';
          strategy:
              'provider_managed_automatic' | 'anthropic_inline_breakpoint' | 'caching_disabled';
      }
    | {
          disposition: 'ignored';
          warning: PromptProviderCacheBoundaryWarning;
      };

export interface PromptProviderCacheBoundaryDto {
    boundary_id: string;
    after_block_id: string;
    after_message_sequence: number | null;
    role_filter: PromptCacheRoleFilterDto;
    ttl: PromptCacheTtl;
    mode: PromptCacheMode;
    disposition: PromptProviderCacheBoundaryDispositionDto;
}

export interface PromptAppliedParameterPreviewDto {
    field: string;
    value_kind: 'null' | 'boolean' | 'number' | 'string' | 'array' | 'object';
    item_count: number | null;
}

export interface PromptDiffEntryDto {
    sequence: number;
    block_id: string;
    requested_role: PromptBlockRoleHint;
    effective_role: 'system' | 'developer' | 'user' | 'assistant';
    wire_role: 'system' | 'developer' | 'user' | 'assistant' | 'model';
    placement: 'message' | 'system_instruction';
}

export type PromptWarningCodeDto =
    | 'cache_boundary_ignored_unsupported'
    | 'cache_boundary_ignored_limit'
    | 'reasoning_effort_omitted'
    | 'creativity_ignored'
    | 'knowledge_disabled'
    | 'memory_disabled'
    | 'cache_reuse_suboptimal'
    | 'missing_module_dependencies'
    | 'resolved_prompt_transform_failed'
    | 'resolved_prompt_transform_ignored'
    | 'provider_cache_boundary_ignored'
    | 'other';

export interface PromptPlanPreviewDto {
    generation_attempt_id: string;
    plan_id: string;
    plan_hash: string;
    prompt_preset_id: string;
    prompt_preset_revision: number;
    generation_target: GenerationTargetDto;
    estimated_input_tokens: number;
    available_input_tokens: number;
    token_estimator_id: string;
    token_estimate_exact: boolean;
    messages: PromptPlanMessagePreviewDto[];
    provider_family: PromptProviderFamily;
    provider_messages: PromptProviderMessagePreviewDto[];
    provider_cache_boundaries: PromptProviderCacheBoundaryDto[];
    cache_directives: PromptCacheDirectivePreviewDto[];
    blocks: PromptBlockResolutionTraceDto[];
    role_mappings: PromptRoleMappingTraceDto[];
    overflow: PromptOverflowTraceDto[];
    warnings: PromptWarningCodeDto[];
    truncated: boolean;
    applied_parameters: PromptAppliedParameterPreviewDto[];
    prompt_diff: PromptDiffEntryDto[];
}

export interface ExplainPromptPlanInput extends Omit<PromptPlanRequestInput, 'expected_plan_hash'> {
    plan_hash: string;
}

export interface PromptPlanRequestInput {
    conversation_id: string;
    branch_id: string;
    expected_head: string | null;
    user_text: string;
    generation_target: GenerationTargetDto;
    prompt_preset_id: string | null;
    variable_overrides: OrchestrationVariableMapDto;
    expected_plan_hash: string | null;
    /** New and resume identities are mutually exclusive; supply exactly one. */
    operation_nonce?: string | null;
    generation_attempt_id?: string | null;
}

/** Reviewed dispatch resumes only the exact preview attempt; it never accepts a nonce. */
export interface ReviewedPromptSendInput extends Omit<
    PromptPlanRequestInput,
    'expected_plan_hash' | 'operation_nonce' | 'generation_attempt_id'
> {
    expected_plan_hash: string;
    generation_attempt_id: string;
}

File: crates/shell-api/src/dto/generation.rs
use chrono::{DateTime, Utc};
use lorepia_core::{
    CHAT_EVENT_VERSION, ChatEvent, ChatEventKind, GenerationTarget, GenerationUsage,
    MessageActionGeneration, MessageRole, RuntimeGenerationCapability, VariableMap,
};
use serde::{Deserialize, Serialize};

use super::{ConversationBranchDto, ConversationModeDto, MessageStatusDto};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case", deny_unknown_fields)]
pub enum GenerationSelectionInput {
    LegacyProfile { provider_profile_id: String },
    Target { target: GenerationTargetDto },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum RuntimePromptRoleInput {
    System,
    User,
    Assistant,
}

impl From<RuntimePromptRoleInput> for MessageRole {
    fn from(value: RuntimePromptRoleInput) -> Self {
        match value {
            RuntimePromptRoleInput::System => Self::System,
            RuntimePromptRoleInput::User => Self::User,
            RuntimePromptRoleInput::Assistant => Self::Assistant,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct RuntimePromptMessageInput {
    pub role: RuntimePromptRoleInput,
    pub content: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum RuntimeGenerationCapabilityInput {
    #[serde(rename = "model:primary")]
    Primary,
    #[serde(rename = "model:auxiliary")]
    Auxiliary,
}

impl From<RuntimeGenerationCapabilityInput> for RuntimeGenerationCapability {
    fn from(value: RuntimeGenerationCapabilityInput) -> Self {
        match value {
            RuntimeGenerationCapabilityInput::Primary => Self::Primary,
            RuntimeGenerationCapabilityInput::Auxiliary => Self::Auxiliary,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct RuntimeGenerationAuditInput {
    pub character_id: String,
    pub character_content_revision_id: Option<String>,
    pub capability: RuntimeGenerationCapabilityInput,
    pub grant_sha256: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct GenerateRuntimeTextInput {
    pub request_id: String,
    pub audit: RuntimeGenerationAuditInput,
    pub selection: GenerationSelectionInput,
    pub messages: Vec<RuntimePromptMessageInput>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct RuntimeTextGenerationDto {
    pub request_id: String,
    pub result: String,
    pub usage: GenerationUsageDto,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct SendMessageInput {
    pub conversation_id: String,
    pub branch_id: String,
    pub expected_head: Option<String>,
    pub mode: ConversationModeDto,
    pub text: String,
    pub selection: GenerationSelectionInput,
    /// Per-generation character/runtime values merged after stored prompt state.
    #[serde(default)]
    pub variable_overrides: VariableMap,
    /// Caller-owned idempotency identity. Missing fields still deserialize so
    /// older clients receive a bounded validation error instead of a schema
    /// decoding failure.
    #[serde(default)]
    pub operation_nonce: Option<String>,
    /// Exact durable attempt to resume. This and `operation_nonce` are XOR.
    #[serde(default)]
    pub generation_attempt_id: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct EditUserMessageInput {
    pub conversation_id: String,
    pub branch_id: String,
    pub expected_head: Option<String>,
    pub message_id: String,
    pub replacement_text: String,
    pub selection: GenerationSelectionInput,
    /// Caller-owned identity for a new edit operation.
    #[serde(default)]
    pub operation_nonce: Option<String>,
    /// Exact durable edit attempt to resume; mutually exclusive with nonce.
    #[serde(default)]
    pub generation_attempt_id: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct RegenerateAssistantMessageInput {
    pub conversation_id: String,
    pub branch_id: String,
    pub expected_head: Option<String>,
    pub message_id: String,
    pub selection: GenerationSelectionInput,
    /// Caller-owned identity for a new regenerate operation.
    #[serde(default)]
    pub operation_nonce: Option<String>,
    /// Exact durable regenerate attempt to resume; mutually exclusive with nonce.
    #[serde(default)]
    pub generation_attempt_id: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct GenerationTargetDto {
    pub model_route_id: String,
    pub generation_preset_id: String,
}

impl From<GenerationTargetDto> for GenerationTarget {
    fn from(value: GenerationTargetDto) -> Self {
        Self {
            model_route_id: value.model_route_id.into(),
            generation_preset_id: value.generation_preset_id.into(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct GenerationStartedDto {
    pub generation_id: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct MessageActionGenerationDto {
    pub branch: ConversationBranchDto,
    pub generation_id: String,
}

impl From<MessageActionGeneration> for MessageActionGenerationDto {
    fn from(value: MessageActionGeneration) -> Self {
        Self {
            branch: value.branch.into(),
            generation_id: value.generation_id.0,
        }
    }
}

/// UI projection of provider-neutral usage counters.
///
/// `provider_raw_summary` is deliberately omitted. It is not required to
/// render chat and remains inside Rust.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct GenerationUsageDto {
    pub input_tokens: Option<u64>,
    pub cached_read_tokens: Option<u64>,
    pub cached_write_tokens: Option<u64>,
    pub output_tokens: Option<u64>,
    pub reasoning_tokens: Option<u64>,
    pub tool_tokens: Option<u64>,
}

impl From<GenerationUsage> for GenerationUsageDto {
    fn from(value: GenerationUsage) -> Self {
        Self {
            input_tokens: value.input_tokens,
            cached_read_tokens: value.cached_read_tokens,
            cached_write_tokens: value.cached_write_tokens,
            output_tokens: value.output_tokens,
            reasoning_tokens: value.reasoning_tokens,
            tool_tokens: value.tool_tokens,
        }
    }
}

/// UI-safe projection of the current Core `ChatEventKind`.
///
/// Variant names and payload topology intentionally match `ChatEvent` v4.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "type", content = "payload", rename_all = "snake_case")]
pub enum ChatEventKindDto {
    GenerationStarted,
    ReasoningDelta(String),
    TextDelta(String),
    ToolCallStarted {
        id: String,
        name: String,
    },
    ToolCallArgumentsDelta {
        id: String,
        delta: String,
    },
    ToolCallCompleted {
        id: String,
    },
    UsageUpdated(GenerationUsageDto),
    MessageCommitted {
        message_id: String,
        status: MessageStatusDto,
    },
    GenerationCancelled,
    GenerationFailed {
        code: String,
        message: String,
    },
    GenerationFinished,
}

impl From<ChatEventKind> for ChatEventKindDto {
    fn from(value: ChatEventKind) -> Self {
        match value {
            ChatEventKind::GenerationStarted => Self::GenerationStarted,
            ChatEventKind::ReasoningDelta(delta) => Self::ReasoningDelta(delta),
            ChatEventKind::TextDelta(delta) => Self::TextDelta(delta),
            ChatEventKind::ToolCallStarted { id, name } => Self::ToolCallStarted {
                id: id.into_inner(),
                name: name.into_inner(),
            },
            ChatEventKind::ToolCallArgumentsDelta { id, delta } => Self::ToolCallArgumentsDelta {
                id: id.into_inner(),
                delta: delta.into_inner(),
            },
            ChatEventKind::ToolCallCompleted { id } => Self::ToolCallCompleted {
                id: id.into_inner(),
            },
            ChatEventKind::UsageUpdated(usage) => Self::UsageUpdated(usage.into()),
            ChatEventKind::MessageCommitted { message_id, status } => Self::MessageCommitted {
                message_id: message_id.0,
                status: status.into(),
            },
            ChatEventKind::GenerationCancelled => Self::GenerationCancelled,
            ChatEventKind::GenerationFailed { code, message } => {
                Self::GenerationFailed { code, message }
            }
            ChatEventKind::GenerationFinished => Self::GenerationFinished,
        }
    }
}

impl ChatEventKindDto {
    pub(crate) const fn is_terminal(&self) -> bool {
        matches!(
            self,
            Self::GenerationCancelled | Self::GenerationFailed { .. } | Self::GenerationFinished
        )
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ChatEventDto {
    pub event_version: u32,
    pub generation_id: String,
    pub conversation_id: String,
    pub branch_id: Option<String>,
    pub assistant_message_id: Option<String>,
    pub sequence: u64,
    pub emitted_at: DateTime<Utc>,
    pub kind: ChatEventKindDto,
}

impl From<ChatEvent> for ChatEventDto {
    fn from(value: ChatEvent) -> Self {
        Self {
            event_version: value.event_version,
            generation_id: value.generation_id.0,
            conversation_id: value.conversation_id.0,
            branch_id: value.branch_id.map(|id| id.0),
            assistant_message_id: value.assistant_message_id.map(|id| id.0),
            sequence: value.sequence,
            emitted_at: value.emitted_at,
            kind: value.kind.into(),
        }
    }
}

impl ChatEventDto {
    pub const fn is_supported_version(&self) -> bool {
        self.event_version == CHAT_EVENT_VERSION
    }
}
